### ISSUE TITLE
[sklearn#33965] Native categorical splitting enhancements in DecisionTree, ExtraTree, RandomForest and ExtraTrees

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#367]
**Mirrored from [scikit-learn#33965](https://github.com/scikit-learn/scikit-learn/issues/33965)**

Original title: Native categorical splitting enhancements in DecisionTree, ExtraTree, RandomForest and ExtraTrees
Original labels: New Feature, Hard, module:tree
Created: 2026-05-10T17:10:50Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

### Describe the workflow you want to enable

From: https://github.com/scikit-learn/scikit-learn/pull/33354#issue-3973551384, we want to enable:

1. [ ] [Optimization] Add a bitset cache (i.e. a uint64_t memory view) that allows us to compute the category split as we traverse the tree for inference speedup
2. [ ] [Feature] Support in tree-based ensemble models (RF + ExtraTrees). This could be done after for random splitter support, so we do all the classes at once
4. [ ] [Refactor] Consolidate the goes_left operation to work at inference time too: https://github.com/scikit-learn/scikit-learn/pull/33972
5. [x] [Refactor] Understand the bitset operations and merge these with the HistGradientBoosting bitset operations https://github.com/scikit-learn/scikit-learn/pull/33354)
6. [ ] [Feature] Implement categorical splitting for RandomSplitter for ExtraTrees. Pretty easy now that we have an arbitrary sized bitset, which we can expand to hold many categories in the case of a randomized tree: https://github.com/scikit-learn/scikit-learn/pull/33972
8. [ ] [Feature] Add support for categorical in exporting 
9. [x] [Feature] Add internal ordinal encoding support for any kind of "categorical input" (see https://github.com/scikit-learn/scikit-learn/pull/33354#discussion_r3199995969)
10. [ ] [Refactor] Address refactoring to use ColumnTransformer in tree, similar to HGBT: https://github.com/scikit-learn/scikit-learn/pull/33354#discussion_r3521665892

Below the line for priority:
1. [ ] [Feature] Brute-force search; expands categorical support to multi-class and multi-output problems
2. [ ] [Feature] Add support for sparse categorical splitting with best and random splitter

### Describe your proposed solution

These work streams may not all be relevant, or mergable, but we can explore.

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to [40](https://github.com/godofecht/flow-scikit/issues/40) in the flow-scikit repository. Further library-specific triage and implementation will happen there.

**Comment by godofecht** [comment]
Moved to [41](https://github.com/godofecht/flow-scikit/issues/41) in the flow-scikit repository. Further library-specific triage and implementation will happen there.
